cep-recipe-serialization.md

```diff
@@ -0,0 +1,1190 @@
+  target_platform: osx-arm64
+  host_platform: osx-arm64
+  build_platform: osx-arm64
+  variant:
+    target_platform: osx-arm64
+  hash:
+    hash: 60d57d3
+    prefix: "" # would contain `py311` and similar prefixes based on variant config
+  directories:
+    host_prefix: # host prefix with placeholders for the host environment
+    build_prefix: # build prefix for the build environment
+    work_dir: # work dir with extracted sources
+    build_dir: # build dir that contains host_prefix, build_prefix and work_dir as subdirectories
+  channels:
+    - conda-forge
+  channel_priority: strict
+  solve_strategy: highest
+  timestamp: 2024-04-13T14:35:30.774863Z
+  subpackages:
+    curl:
+      name: curl
+      version: 8.0.1
+      build_string: h60d57d3_0
+  packaging_settings:
+    archive_type: conda
+    compression_level: 15
+```
+
+### Finalized dependencies
+
+The finalized dependencies section contains a "lockfile" of all the dependencies
+exactly as they were used at build-time, for the two environments `build` and
+`host`. It also includes the final `run` and `constraints` specs that are
+computed based on the `build` and `host` dependencies (resolved all
+`run_exports`).
+
+The `finalized_dependencies` section contains the following keys:
+
+```yaml
+finalized_dependencies:
+  build:
+    specs: [list of spec objects]
+    resolved: [list of resolved dependency objects]
+    run_exports: { package_name: run_export } # all run_exports.json that were found among the packages in this environment
+  host:
+    specs: [list of spec objects]
+    resolved: [list of resolved dependency objects]
+    run_exports: { package_name: run_export } # all run_exports.json that were found among the packages in this environment
+  run:
+    depends: [list of dependencies with dependency source information]
+    constraints: [list of constraints with dependency source information]
+    run_exports: optional run_export dictionary
+```
+
+#### Spec objects
+
+A dependency spec contains the `MatchSpec` and additional information about the
+source of the dependency. There are 6 possible `DependencyInfo` versions:
+
+- `pin_subpackage`: The dependency comes from a `pin_subpackage` expression.
+- `pin_compatible`: The dependency comes from a `pin_compatible` expression.
+- `source`: The dependency comes from te recipe.
+- `run_export`: The dependency comes from a run export.
+- `variant`: The dependency comes from a variant.
+
+Example:
+
+```yaml
+# language specified in ${{ compiler(`c`) }}
+- source: clang_osx-arm64   
+- source: make >=1.3
+- pin_compatible: quarto
+  lower_bound: x.x    # not present if set to None
+  upper_bound: x.x.x  # not present if set to None
+  exact: true         # not present if false
+  spec: quarto >=1.4.550,<1.5
+- pin_subpackage: some-pkg-a
+  lower_bound: x.x    # not present if set to None
+  upper_bound: x.x.x  # not present if set to None
+  exact: true         # not present if false
+  spec: some-pkg-a >=1.0.0,<2
```

Comment:
Am curious how these work together. We have `exact` and a more relaxed `spec` as well bounds. What takes precedence?

Reply:
They should not be allowed together, right?